src/core/event/group.cpp

```diff
@@ -0,0 +1,113 @@
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "config.h"
+#include "group.h"
+
+#include "dev/ring.h"
+#include "event/event_handler_manager_local.h"
+#include "sock/sockinfo_tcp.h"
+
+#define MODULE_NAME "group:"
+
+#define grp_logpanic __log_panic
+#define grp_logerr   __log_err
+#define grp_logwarn  __log_warn
+#define grp_loginfo  __log_info
+#define grp_logdbg   __log_dbg
+
+xlio_group::xlio_group(const struct xlio_group_attr *attr)
+    : m_socket_event_cb(attr->socket_event_cb)
+    , m_socket_comp_cb(attr->socket_comp_cb)
+    , m_socket_rx_cb(attr->socket_rx_cb)
+    , m_flags(attr->flags)
+{
+    m_rings.reserve(2);
+    m_event_handler = new event_handler_manager_local();
+    m_tcp_timers = new tcp_timers_collection(safe_mce_sys().tcp_timer_resolution_msec,
+                                             safe_mce_sys().tcp_timer_resolution_msec);
+    m_tcp_timers->set_group(this);
+}
+
+xlio_group::~xlio_group()
+{
+    delete m_tcp_timers;
+    delete m_event_handler;
+}
+
+void xlio_group::poll()
+{
+    for (ring *rng : m_rings) {
+        uint64_t sn;
+        rng->poll_and_process_element_tx(&sn);
+        sn = 0;
+        rng->poll_and_process_element_rx(&sn);
+    }
+    m_event_handler->do_tasks();
+}
+
+void xlio_group::add_dirty_socket(sockinfo_tcp *si)
+{
+    if (m_flags & XLIO_GROUP_FLAG_DIRTY) {
+        m_dirty_sockets.push_back(si);
+    }
+}
+
+void xlio_group::flush()
+{
+    for (auto si : m_dirty_sockets) {
+        si->flush();
+    }
+    m_dirty_sockets.clear();
+    // TODO Ring doorbell and request TX completion.
+}
+
+void xlio_group::add_ring(ring *rng)
+{
+    if (std::find(m_rings.begin(), m_rings.end(), rng) == std::end(m_rings)) {
+        grp_loginfo("New ring %p in group %p", rng, this);
+        rng->set_group(this);
```

Comment:
I wonder if it worth checking if a ring already belongs to a group ....

Reply:
Probably as sanity check or assert. If we see a ring here, that means the socket uses special ring allocation logic with the key equals to the group pointer. Therefore, in GA version this shouldn't be possible.

Current implementation can lead to a situation when a group is destroyed without destroying the sockets and then new group is allocated with the same address - the old ring will be reused for the new group. Even such a corner case should work functionally. GA version will terminate (reset) the remaining sockets on a group destruction, so the ring will be released.

There is AI for GA version to cover the above. Will add a sanity check for current version that a group is not set.

Reply:
Done. Added warning in case of a group is already set.